design/ErrorHandling.md

```diff
@@ -0,0 +1,160 @@
+# Error Handling
+
+The simplest design for error handling would be to do it synchronously, for example with Javascript exceptions and object creation returning `null`.
+However, this would introduce a lot of synchronization points for multi-threaded/multi-process WebGPU implementations, making too slow to be useful.
+
+There are a number of cases that developers or applications need error handling for:
+
+ - *Debugging*: Getting errors synchronously during development, to break in to the debugger.
+ - *Telemetry*: Collecting error logs in deployment, for bug reporting and telemetry.
+ - *Recovery*: Recovering from recoverable errors (like out-of-memory on resource creation).
+ - *Fallback*: Tearing down the application and falling back, e.g. to WebGL, 2D Canvas, or static content.
+
+Meanwhile, error handling should not make the API clunky to use.
+
+There are several types of WebGPU calls that get their error handled differently:
+
+ - *Creation*: WebGPU Object creation.
+ - *Encoding*: Recording of GPU commands in `WebGPUCommandEncoder`.
+ - *Operations*: Other commands, such as `WebGPUQueue.submit`.
+
+## *Telemetry*: Error Logging
+
+`WebGPUDevice` keeps a log of errors that happened on each WebWorker which the application can query asynchronously.
+These errors are not directly associated with the objects or operations that produced them.
+They should not be used by applications to recover from expected errors.
+Instead, they are used for handling unexpected errors in deployment, for bug reporting and telemetry.
+
+```
+enum WebGPULogEntryType {
+    "device-lost",
+    "validation-error",
+    "recoverable-out-of-memory",
+};
+
+interface WebGPULogEntry {
+    readonly attribute WebGPULogEntryType type;
+    readonly attribute DOMString? reason;
+};
+
+partial interface WebGPUDevice {
+    Promise<sequence<WebGPULogEntry>> getCurrentErrorLog();
+};
```

Comment:
Nit: Since promises are one-shot objects, the document should clarify that getCurrentErrorLog returns a brand new promise every time it is called.